### ISSUE TITLE
feof(): Argument #1 ($stream) must be of type resource, null given after upgrading to laravel 9 for files uploaded before the upgrade

### ISSUE THREAD

**Issue by ziming** [spatie/laravel-medialibrary#2842]
So I have a route that trigger a MediaStream download. Returning a zip file with multiple media in it.

Timeline of events:
Many days before upgrading to laravel 9 i upgraded to PHP 8.1 for all my sites, so unlikely a language version issue.

After upgrading to laravel 9 for my 1st small site (4th March). I noticed the following issues on production:
- Single file download still works fine
- MediaStream downloads starts throwing `feof(): Argument #1 ($stream) must be of type resource, null given` errors. However it only happens if the MediaStream contains a media/file that is uploaded before 4th March 2022.
- If all the files in the MediaStream is uploaded after 4th March 2022. The MediaStream downloads works fine. Thankfully there isn't a lot of files in the 1st production site. So we just do with downloading the files individually for the older files.

On 15 March 2022. We upgraded our 2nd site to laravel 9 too. Same issues again. If the MediaStream contain files uploaded after 15 March. the zip file download work fine. But if the MediaStream contain files uploaded before 15 March. The exception "`feof(): Argument #1 ($stream) must be of type resource, null given`" is thrown.

In my error logs, the stack trace are as follows:

Line 157: vendor/maennchen/zipstream-php/src/Stream.php
Line 353: vendor/maennchen/zipstream-php/src/File.php
Line 407: vendor/maennchen/zipstream-php/src/File.php
Line 336: vendor/maennchen/zipstream-php/src/File.php
Line 296: vendor/maennchen/zipstream-php/src/ZipStream.php
Line 82: vendor/spatie/laravel-medialibrary/src/Support/MediaStream.php

Given these 2 sites MediaStream download problem only come up after the laravel 9 upgrade. I conclude that it is unlikely an issue on AWS side too.

So I went to check the commit history of MediaStream.php. I don't see anything that could break backward compatibility for older uploads given that the only commits from Aug 27 2020 to 14 Jan 2022 are just simple code changes that uses PHP 8 features. Mainly type hintings & constructor promotions. As well as a change from new static to new self in the create method.

https://github.com/spatie/laravel-medialibrary/blob/main/src/Support/MediaStream.php

Then I went to check the Zipstream-PHP package. The last commit is on 2 Nov 2021. Which makes it unlikely to be caused by a recent change in that package too. Since we are using the latest version of that package even in laravel 8 & medialibrary 9

https://github.com/maennchen/ZipStream-PHP

So now that brings me to https://github.com/thephpleague/flysystem-aws-s3-v3

My hunch tell me that since we jumped from flysystem-aws-s3-v1 to v3 from laravel 8 to laravel 9. Something changed in that package is likely the cause, maybe the way files are uploaded to S3 or something else. I'm not sure which. I decided to post this issue here instead of flysystem-aws-s3-v3 because this is the place where most people will likely encounter the same problem. If @frankdejonge see this & knows the reason that would be great too.

So if anyone could investigate further it will be great.

TLDR:
- After upgrade to laravel 9. MediaStream throws an exception if it include files that are uploaded to S3 before the upgrade to laravel 9. If the files were uploaded after the laravel 9 upgrade. MediaStream download proceeds without problems.
- Single file downloads continues to work fine.

**Comment by freekmurze** [comment]
Thanks for having researched this. I'd accept a PR that solves this.

**Comment by ziming** [comment]
I've decided to login to my AWS account to look at the S3 buckets and I noticed the following in both sites separate S3 buckets.

![image](https://user-images.githubusercontent.com/679513/158628566-5aec0474-0b43-4f12-b81a-0b8e5d9c5351.png)

It appears that when a user uploads a file in laravel 9. it's uploaded with "/" as the top level directory which didn't happen in laravel 8

in terms of config. (they stayed the same in laravel 8 and 9)

for my config/media-library.php

```php
'prefix' => env('MEDIA_PREFIX', ''),
```

for my config/fileystems.php

```php
's3_uploads' => [
            'driver' => 's3',
            'root' => '/uploads',
            'key' => env('AWS_ACCESS_KEY_ID'),
            'secret' => env('AWS_SECRET_ACCESS_KEY'),
            'region' => env('AWS_DEFAULT_REGION'),
            'bucket' => env('AWS_BUCKET'),
            'url' => env('AWS_URL'),
        ],
```
